lib/propolis/src/cpuid.rs

```diff
@@ -0,0 +1,413 @@
+                    // is stored in bits 25:14
+                    vals.eax &= !(0xfff << 14);
+                    vals.eax |= (visible_count - 1) << 14;
+                }
+            }
+        }
+    }
+    fn fix_cpu_topo(&self, set: &mut Set) {
+        for topo in enum_iterator::all::<TopoKind>() {
+            // Nuke any existing info in order to potentially override it
+            let leaf = topo as u32;
+            set.remove_all(leaf);
+
+            let num_vcpu = match self.num_vcpu {
+                Some(n) => n.get() as u32,
+                None => {
+                    // Skip adding topo info if we lack vcpu count
+                    continue;
+                }
+            };
+
+            if self.cpu_topo.contains(&topo) {
+                match topo {
+                    TopoKind::StdB => {
+                        // Queries with invalid ecx will get all-zeroes
+                        set.insert(Ident(leaf, None), Entry::zero());
+                        if self.has_smt {
+                            set.insert(
+                                Ident(leaf, Some(0)),
+                                Entry {
+                                    eax: 0x1,
+                                    ebx: 0x2,
+                                    ecx: 0x100,
+                                    // TODO: populate with x2APIC ID
+                                    edx: 0x0,
+                                },
+                            );
+                        } else {
+                            set.insert(
+                                Ident(leaf, Some(0)),
+                                Entry {
+                                    eax: 0x0,
+                                    ebx: 0x1,
+                                    ecx: 0x100,
+                                    // TODO: populate with x2APIC ID
+                                    edx: 0x0,
+                                },
+                            );
+                        }
+                        set.insert(
+                            Ident(leaf, Some(1)),
+                            Entry {
+                                eax: 0x0,
+                                ebx: num_vcpu,
+                                ecx: 0x201,
+                                // TODO: populate with x2APIC ID
+                                edx: 0x0,
+                            },
+                        );
+                    }
+                    TopoKind::Std1F => {
+                        // TODO: add 0x1f topo info
+                    }
+                    TopoKind::Ext1E => {
+                        let id = self.vcpuid.unwrap_or(0) as u32;
+                        let compute_id = if self.has_smt { id / 2 } else { id };
+                        set.insert(
+                            Ident(leaf, None),
+                            Entry {
+                                eax: id,
+                                ebx: 0x100 | id,
```

Comment:
If I'm reading things right, `ebx` bit 8 should be cleared if `has_smt` is false.

Reply:
Yeah